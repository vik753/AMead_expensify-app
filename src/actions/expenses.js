import database from '../firebase/firebase';

// ADD_EXPENSE
const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0,
        } = expenseData;
        const expense = {description, note, amount, createdAt};

        return database.ref('expenses')
            .push(expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }));
            }).catch((e) => console.log('Error pushing new expense: ', e));
    }
};
// REMOVE_EXPENSE
const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

const startRemoveExpense = ({id} = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove()
            .then(() => {
                dispatch(removeExpense({id}));
                // console.log('Remove succeeded!');
            })
            .catch((e) => console.log('Error: ', e.message));
    };
};
// EDIT_EXPENSE
const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});
// SET_EXPENSES
const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses')
            .once('value')
            .then((snapshot) => {
                const expenses = [];
                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });
                dispatch(setExpenses(expenses));
            }).catch(e => console.log('Error loading expenses: ', e));
    };
};

export {
    startSetExpenses,
    startAddExpense,
    startRemoveExpense,
    addExpense,
    removeExpense,
    editExpense,
    setExpenses,
}