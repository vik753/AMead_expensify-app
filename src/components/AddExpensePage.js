import React from 'react';

const AddExpensePage = () => (
    <div>This is  AddExpansePage</div>
);

export { AddExpensePage }