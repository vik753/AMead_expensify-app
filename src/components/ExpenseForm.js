import React from 'react';
import moment from "moment";
import {SingleDatePicker} from "react-dates";
import 'react-dates/lib/css/_datepicker.css';

// const date = new Date();
const now = moment();
console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            id: props.expense ? props.expense.id : undefined,
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) :  moment(),
            calendarFocused: false,
            error: '',
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}));
        }
    };
    onDateChange = (createdAt) => {
        createdAt && this.setState({createdAt: createdAt})
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description && !this.state.amount) {
            this.setState(() => ({error: 'Please provide the description and amount!'}))
        } else {
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                id: this.state.id,
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note,
            })
        }
    };

    render() {
        return (
            <div>
                {this.state.error && (<p><b>{this.state.error}</b></p>)}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt} // momentPropTypes.momentObj or null
                        onDateChange={this.onDateChange} // PropTypes.func.isRequired
                        focused={this.state.calendarFocused} // PropTypes.bool
                        onFocusChange={({focused}) => this.setState({calendarFocused: focused})} // PropTypes.func.isRequired
                        numberOfMonths={1} // how months will be open in datepicker
                        firstDayOfWeek={1} // 1 - Monday
                        isOutsideRange={() => false} // lets us to choice the past dates
                        displayFormat={'DD.MM.YYYY'} // showing date format
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>{this.state.id ? 'Update expense' : 'Add Expense'}</button>
                </form>
            </div>
        );
    }
}