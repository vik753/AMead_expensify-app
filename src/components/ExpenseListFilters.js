import React from 'react';
import {connect} from "react-redux";
import {DateRangePicker, SingleDatePicker} from "react-dates";
import {
    setTextFilter,
    sortByDate,
    sortByAmount,
    setStartDate,
    setEndDate
} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };
    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState({
            calendarFocused
        });
    };
    onTextChange = (e) => {
        const textFilter = e.target.value;
        this.props.setTextFilter(textFilter);
    };
    onSortChange = (e) => {
        const sortFilter = e.target.value;
        sortFilter === 'date' && this.props.sortByDate();
        sortFilter === 'amount' && this.props.sortByAmount();
    };

    render() {
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                    //startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                    endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                    //endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                    onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                    focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                    numberOfMonths={1} // how months will be open in datepicker
                    firstDayOfWeek={1} // 1 - Monday
                    isOutsideRange={() => false} // lets us to choice the past dates
                    displayFormat={'DD.MM.YYYY'} // showing date format
                    showClearDates={true}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (textFilter) => dispatch(setTextFilter(textFilter)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);