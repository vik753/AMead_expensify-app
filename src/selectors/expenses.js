// Get visible expenses
import moment from "moment";

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        // console.log(startDateMatch, endDateMatch);
        return startDateMatch && endDateMatch && textMatch;
    }).sort((expenseA, expenseB) => {
        if (sortBy === 'date') {
            return expenseA.createdAt < expenseB.createdAt ? 1 : -1;
        }
        if (sortBy === 'amount') {
            return expenseA.amount < expenseB.amount ? 1 : -1;
        }
    });
};

export {getVisibleExpenses}