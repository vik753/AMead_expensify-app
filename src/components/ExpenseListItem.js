import React, {Component} from 'react';
import moment from "moment";
import {connect} from "react-redux";
import {Link} from "react-router-dom";
import {addExpense, removeExpense, editExpense} from '../actions/expenses';

export class ExpenseListItem extends Component {

    handleRemoveBtn = () => {
        const {dispatch, id} = this.props;
        dispatch(removeExpense({id}));
    };

    render() {
        const {id, amount, createdAt, description, note} = this.props;
        // console.log(this.props);
        return (
            <li className='listItem'
                id={id}
            >
                <Link to={`/edit/${id}`}>
                    <span>{description} </span>
                </Link>
                <span> ${parseFloat((amount / 100)).toFixed(2)}, </span>
                <span>{moment(createdAt).format('MMM Do, YYYY')} </span>
                <span>note: {note} </span>
                <button
                    onClick={this.handleRemoveBtn}
                > Remove
                </button>
            </li>
        );
    }
}

const mapStateToProps = (state) => ({
    expenses: state.expenses,
});

export default connect(mapStateToProps)(ExpenseListItem);

/*
* amount: 325
createdAt: 1922
description: "Water bill"
id: "44b464bd-e7c2-4b6f-bdb9-74bf49808a67"
note: "Water bill"
* */