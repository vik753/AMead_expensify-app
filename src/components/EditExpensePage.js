import React from 'react';
import {connect} from "react-redux";
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from '../actions/expenses';


const EditExpensePage = (props) => {
    // console.log(props.match.params.id);
    return (
        //props.match.params.id
        <div>
            <ExpenseForm
                expense={props.expense}
                onSubmit={(expense) => {
                    // console.log(props.expense.id, expense);
                    props.dispatch(editExpense(expense.id, expense));
                    props.history.push('/');
                }}
            />
            <button
                onClick={() => {
                    props.dispatch(removeExpense({id: props.expense.id}));
                    props.history.push('/');
                }}
            >
                Remove
            </button>
        </div>
    );
};

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),
});

export default connect(mapStateToProps)(EditExpensePage);