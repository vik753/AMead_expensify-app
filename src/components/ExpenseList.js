import React from 'react';
import {connect} from 'react-redux';
import getVisibleExpenses from '../selectors/expenses';
import ExpenseListItem from "./ExpenseListItem";

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <h1>No Expenses</h1>
            ) : (
                <div>
                    <h1>Expense list</h1>
                    <hr/>
                    <ol>
                        {
                            props.expenses.map((expense) => {
                                return (<ExpenseListItem key={expense.id} {...expense} />)
                            })
                        }
                    </ol>
                </div>
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters),
    }
};

export default connect(mapStateToProps)(ExpenseList);


/*
amount: 325
createdAt: 1922
description: "Water bill"
id: "137e5855-63a9-4d24-832c-71a3a068ed23"
note: "Water bill"*/
