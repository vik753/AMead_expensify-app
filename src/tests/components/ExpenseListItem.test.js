import React from 'react';
import {shallow} from 'enzyme';
import toJson from "enzyme-to-json";
import {ExpenseListItem} from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";

test('should render ExpenseListItem with expenses data', () => {
    const wrapper = shallow(
        <ExpenseListItem
            key={expenses[0].id}
            {...expenses[0]}
        />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
});