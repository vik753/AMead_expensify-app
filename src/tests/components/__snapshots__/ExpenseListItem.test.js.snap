// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem correctly 1`] = `
ShallowWrapper {
  "length": 1,
}
`;
